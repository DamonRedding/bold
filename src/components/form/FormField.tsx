import * as React from 'react'
import * as classnames from 'classnames'
import { ControlProps, Control } from './Control'
import { FormLabelProps, FormLabel } from './FormLabel'

export interface FormFieldProps extends ControlProps, FormLabelProps {
    name?: string
    hasAddons?: boolean
    hasControl?: boolean
    unit?: string
    title?: string
}

export class FormField extends React.Component<FormFieldProps, any> {

    static defaultProps: FormFieldProps = {
        hasControl: true
    }

    render() {
        const fieldClasses = classnames('field is-vertical', this.props.className, {
            'has-error': this.props.error,
        })

        const internalFieldClasses = classnames('field', {
            'is-grouped': this.props.unit !== undefined,
            'has-addons': this.props.hasAddons
        })

        const { hasControl, ...rest } = this.props

        return (
            <div title={this.props.title} data-name={this.props.name} className={fieldClasses}>
                {this.props.label && <FormLabel {...rest } />}
                <div className={internalFieldClasses}>
                    {hasControl ? <Control {...rest} /> :
                        this.props.children}
                    {this.props.unit && <p className='control'><span className='unit'>{this.props.unit}</span></p>}
                </div>
            </div>
        )
    }

}
