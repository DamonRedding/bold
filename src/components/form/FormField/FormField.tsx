import * as React from 'react'

import { FormError, FormErrorProps } from '../FormError'
import { FormLabel, FormLabelProps } from '../FormLabel'

export interface FormFieldProps {
    label?: FormLabelProps['label']
    required?: FormLabelProps['required']
    error?: FormErrorProps['error']
    name?: string
}

export class FormField extends React.Component<FormFieldProps> {

    render() {
        const { children, name, error, label, required } = this.props
        const styles = {
            label: {
                display: 'block',
                marginBottom: '0.25rem',
            },
            error: {
                marginTop: '0.25rem',
            },
        }

        return (
            <div data-name={name}>
                {label &&
                    <FormLabel required={required} style={styles.label} htmlFor={name} label={label} />
                }
                {children}
                {error &&
                    <FormError style={styles.error} error={error} />
                }
            </div>
        )
    }

}
