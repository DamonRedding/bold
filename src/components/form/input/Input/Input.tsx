import * as React from 'react'

export interface InputController {
  getInput(): HTMLInputElement
  focus(): any
  blur(): any
}

export interface PublicInputProps {
  className?: string
  checked?: boolean
  disabled?: boolean
  id?: string
  maxLength?: number
  name?: string
  onBlur?: <T>(event?: React.FocusEvent<T>) => void
  onChange?: <T>(event: React.ChangeEvent<T> | any) => void
  onFocus?: <T>(event?: React.FocusEvent<T>) => void
  onKeyPress?: (event: React.KeyboardEvent<HTMLInputElement>) => void
  placeholder?: string
  value?: any
  provideController?: (controller: InputController) => void
  autoFocus?: boolean
}

export interface InputProps extends PublicInputProps {
  type?: string
}

export class Input extends React.Component<InputProps, any> {
  private input: HTMLInputElement

  componentDidMount() {
    this.props.provideController && this.props.provideController({
      getInput: () => this.input,
      focus: () => this.input.focus(),
      blur: () => this.input.blur(),
    })
  }

  componentWillUnmount() {
    this.props.provideController && this.props.provideController(null)
  }

  render() {
    const { provideController, ...rest } = this.props

    return (
      <input ref={this.ref} {...rest} />
    )
  }

  private ref = (elem: HTMLInputElement) => {
    this.input = elem
  }

}
