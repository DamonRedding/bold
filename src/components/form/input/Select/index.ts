export * from './SelectMenu'
export * from './Select'
