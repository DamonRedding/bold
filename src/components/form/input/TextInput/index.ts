export { TextInputBase, TextInputBaseProps, InputStatus } from './TextInputBase'
export { TextInput, TextInputProps } from './TextInput'
