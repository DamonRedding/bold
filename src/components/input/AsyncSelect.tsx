import * as React from 'react'
import * as Select from 'react-select'
import axios from 'axios'

export interface AsyncSelectProps extends Partial<Select.ReactAsyncSelectProps> {
    url: string
    value?: any
    pageSize?: number
    searchDelay?: number
    customizeParams?: any,
    maxLength?: number
}

export class AsyncSelect extends React.Component<AsyncSelectProps> {
    private typingTimer: number

    static defaultProps: Partial<AsyncSelectProps> = {
        autoload: false,
        pageSize: 10,
        searchDelay: 500,
        valueKey: 'id',
        labelKey: 'nome',
        multi: false,
        cache: false,
        placeholder: '',
        clearValueText: 'Limpar seleção',
        loadingPlaceholder: 'Carregando...',
        searchPromptText: 'Digite para pesquisar',
        noResultsText: 'Nenhum item encontrado.',
        backspaceRemoves: false,
        ignoreAccents: false,
        ignoreCase: true,
        clearable: true,
    }

    render() {
        const { value, onBlur, ...rest } = this.props
        return (
            <Select.Async
                options={[]}
                {...rest}
                inputProps={{ maxLength: this.props.maxLength }}
                value={value && value.toJS ? value.toJS() : value}
                onBlur={this.blur()}
                loadOptions={this.loadOptions}
            />
        )
    }

    private loadOptions = (query, callback) => {
        clearTimeout(this.typingTimer)
        if (this.typingTimer) {
            clearTimeout(this.typingTimer)
        }
        this.typingTimer = setTimeout(this.getPage.bind(this, query, callback), this.props.searchDelay)
    }

    private getPage(query, callback) {
        let params = {
            query: query,
            pageSize: this.props.pageSize
        }

        params = this.props.customizeParams ? this.props.customizeParams(params) : params

        return axios.get(this.props.url, { params: params })
            .then(data => {
                let response = {
                    options: data.data
                }
                callback(null, response)
            }).catch(error => {
                callback(error, null)
            })
    }

    private blur = (): React.EventHandler<React.FocusEvent<{}>> => {
        return () => {
            if (this.props.onBlur) {
                if (this.props.value && this.props.value[this.props.labelKey]) {
                    this.props.onBlur(this.props.value)
                } else {
                    this.props.onBlur(null)
                }
            }
        }
    }
}
