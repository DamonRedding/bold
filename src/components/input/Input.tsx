import * as React from 'react'
import { EventOrValueHandler } from 'redux-form'
import * as classnames from 'classnames'
import { helpersClassnames, extractProps } from '../../util/Util'
import { ChangeEvent, FocusEvent } from 'react'

export interface PublicInputProps {
  className?: string
  checked?: boolean
  disabled?: boolean
  id?: string
  maxLength?: number
  name?: string
  onBlur?: EventOrValueHandler<FocusEvent<any>>
  onChange?: EventOrValueHandler<ChangeEvent<any>>
  onKeyPress?: (event: React.KeyboardEvent<HTMLInputElement>) => void
  placeholder?: string
  value?: any
}

export interface InputProps extends PublicInputProps {
  type: string
}

const propsToKeep = [
  'checked',
  'disabled',
  'id',
  'maxLength',
  'name',
  'onBlur',
  'onChange',
  'onKeyPress',
  'placeholder',
  'type',
  'value'
]

export class Input extends React.Component<InputProps, any> {
  public input: HTMLInputElement

  private ref = (elem: HTMLInputElement) => {
    this.input = elem
  }

  focus() {
    this.input.focus()
  }

  blur() {
    this.input.blur()
  }

  render() {
    const classes = classnames('input', this.props.className, helpersClassnames(this.props))

    return (
      <input ref={this.ref} {...extractProps(this.props, ...propsToKeep) } className={classes} />
    )
  }

}
