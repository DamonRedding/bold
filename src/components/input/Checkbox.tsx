import * as React from 'react'
import * as classnames from 'classnames'
import { WrappedFieldInputProps } from 'redux-form'
import { Input } from './Input'
import { Icon } from '../elements/Icon'


export interface CheckboxProps extends Partial<WrappedFieldInputProps> {
    checked?: boolean
    disabled?: boolean
    label?: string
}

export class Checkbox extends React.Component<CheckboxProps, any> {

    render() {
        const classes = classnames('checkbox', {
            'disabled': this.props.disabled
        })
        const { label, ...rest } = this.props
        return (
            <label className={classes} >
                <Input {...rest} type='checkbox' />
                <Icon className='check-indicator' icon='checkbox-check' />
                <span className='check-label'>{this.props.label || this.props.children}</span>
            </label>
        )
    }

}
