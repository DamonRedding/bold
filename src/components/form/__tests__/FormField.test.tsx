import * as React from 'react'
import { shallowRenderAndMatch } from '../../../__tests__/test.defaults'
import { FormField } from '../FormField'
import { Input } from '../../input/Input'

describe('FormField', () => {
    it('deve renderizar corretamente', function () {
        shallowRenderAndMatch(<FormField />)
    })
    it('deve renderizar o ícone', function () {
        shallowRenderAndMatch(<FormField icon='login' />)
    })
    it('deve renderizar o ícone na direita', function () {
        shallowRenderAndMatch(<FormField icon='login' iconRight />)
    })
    it('deve renderizar o label', function () {
        shallowRenderAndMatch(<FormField label='label teste' />)
    })
    it('deve renderizar o label com o símbolo de required', function () {
        shallowRenderAndMatch(<FormField label='label teste' required />)
    })
    it('deve renderizar a unidade', function () {
        shallowRenderAndMatch(<FormField unit='unidade' />)
    })
    it('deve renderizar os filhos', function () {
        shallowRenderAndMatch(<FormField><Input type='text' /></FormField>)
    })
})
