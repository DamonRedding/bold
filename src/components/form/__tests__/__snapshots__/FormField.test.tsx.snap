// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormField deve renderizar a unidade 1`] = `
<div
  className="field is-vertical"
>
  <div
    className="field is-grouped"
  >
    <Control
      unit="unidade"
    />
    <p
      className="control"
    >
      <span
        className="unit"
      >
        unidade
      </span>
    </p>
  </div>
</div>
`;

exports[`FormField deve renderizar corretamente 1`] = `
<div
  className="field is-vertical"
>
  <div
    className="field"
  >
    <Control />
  </div>
</div>
`;

exports[`FormField deve renderizar o label 1`] = `
<div
  className="field is-vertical"
>
  <FormLabel
    label="label teste"
  />
  <div
    className="field"
  >
    <Control
      label="label teste"
    />
  </div>
</div>
`;

exports[`FormField deve renderizar o label com o símbolo de required 1`] = `
<div
  className="field is-vertical"
>
  <FormLabel
    label="label teste"
    required={true}
  />
  <div
    className="field"
  >
    <Control
      label="label teste"
      required={true}
    />
  </div>
</div>
`;

exports[`FormField deve renderizar o ícone 1`] = `
<div
  className="field is-vertical"
>
  <div
    className="field"
  >
    <Control
      icon="login"
    />
  </div>
</div>
`;

exports[`FormField deve renderizar o ícone na direita 1`] = `
<div
  className="field is-vertical"
>
  <div
    className="field"
  >
    <Control
      icon="login"
      iconRight={true}
    />
  </div>
</div>
`;

exports[`FormField deve renderizar os filhos 1`] = `
<div
  className="field is-vertical"
>
  <div
    className="field"
  >
    <Control>
      <Input
        type="text"
      />
    </Control>
  </div>
</div>
`;
